# Exploratory Data Analysis: WebMD Reviews of Type 2 Diabetes Treatments

## Introduction:
Project Description:
[Enter text here]

### Contents:
 - Read in scraped data
 - Perform QA on data from scrape
 - Exploratory Data Analysis
 - Insights

#### Read in scraped data:
First, let's import numpy and pandas, and read in the scraped data.

In [1]:
import numpy as np
import pandas as pd

In [2]:
reviews = pd.read_csv('webmd_reviews.csv')

Take a look at some sample data

In [3]:
reviews.sample(10)

Rdate  \
4823   10/30/2013 6:14:09 PM   
584     2/6/2008 12:41:44 PM   
3229    2/26/2008 3:35:16 PM   
3748   12/24/2012 2:07:01 PM   
1397     1/7/2009 9:52:52 PM   
1611    9/24/2009 3:58:57 AM   
522    2/26/2009 11:56:36 AM   
573   12/13/2007 12:52:36 AM   
4517    5/5/2010 12:41:15 PM   
2455   3/23/2012 10:38:58 AM   

                                                comment  \
4823                                                NaN   
584                                                 NaN   
3229                                                NaN   
3748                                                NaN   
1397                                                NaN   
1611                                                NaN   
522                                                 NaN   
573   I have taken this medication for the last 10 y...   
4517                                                NaN   
2455                                                NaN   

                                      condition                          drug  \
4823                   Type 2 Diabetes Mellitus                 Invokana oral   
584                    Type 2 Diabetes Mellitus                  Avandia oral   
3229                   Type 2 Diabetes Mellitus                    Actos oral   
3748                   Type 2 Diabetes Mellitus                    Actos oral   
1397                   Type 2 Diabetes Mellitus                metformin oral   
1611                   Type 2 Diabetes Mellitus                metformin oral   
522                    Type 2 Diabetes Mellitus  Lantus Solostar subcutaneous   
573                    Type 2 Diabetes Mellitus                  Avandia oral   
4517  Additional Medication for Diabetes Type 2           Byetta subcutaneous   
2455                   Type 2 Diabetes Mellitus              nateglinide oral   

      easeofuse  effectiveness  helpful  \
4823          5              4        1   
584           5              4        6   
3229          5              4        1   
3748          5              5        1   
1397          2              2        2   
1611          4              2        5   
522           5              3        9   
573           4              3        2   
4517          4              5        6   
2455          5              4        5   

                                               reviewer  satisfaction  
4823  65-74 Female  on Treatment for less than 1 mon...             4  
584   mikeoco, 65-74 Male  on Treatment for 2 to les...             3  
3229  chowdawgpup65, 65-74 Male  on Treatment for 1 ...             4  
3748  amadio, 75 or over Male  on Treatment for 10 y...             5  
1397  xring, 55-64 Male  on Treatment for less than ...             1  
1611  bramaw, 55-64 Female  on Treatment for 1 to 6 ...             1  
522   tndsteward, 35-44 Female  on Treatment for les...             3  
573   Victim of Avandia, 55-64 Female  on Treatment ...             1  
4517  karenb, 65-74 Female  on Treatment for 2 to le...             5  
2455                            65-74 Female  (Patient)             2

In [4]:
reviews.columns

Index(['Rdate', 'comment', 'condition', 'drug', 'easeofuse', 'effectiveness',
       'helpful', 'reviewer', 'satisfaction'],
      dtype='object')

#### Perform Basic QA

Now let's check against WebMD to ensure counts and classifications were scraped as expected.

Reference URL: https://www.webmd.com/drugs/2/condition-594/type%202%20diabetes%20mellitus

How many reviews did we pull?

In [5]:
len(reviews.drug)

5495

How many unique drugs with reviews did we scrape?

In [6]:
reviews['drug'].nunique()

70

How many reviews from each drug did we scrape?

In [7]:
reviews['drug'].value_counts()

metformin oral                               1227
Actos oral                                    622
Januvia oral                                  440
Byetta subcutaneous                           373
Janumet oral                                  208
glipizide oral                                195
Invokana oral                                 193
glimepiride oral                              172
Trulicity subcutaneous                        126
Glucophage oral                               111
Bydureon subcutaneous                         110
Onglyza oral                                  106
glyburide oral                                101
Avandia oral                                   96
Amaryl oral                                    95
Tradjenta oral                                 90
Farxiga oral                                   72
Actoplus MET oral                              68
Kombiglyze XR oral                             58
Levemir Flexpen subcutaneous                   56


We searched for "Type 2 Diabetes Mellitus" on WebMD, but are these the only conditions?  Let's check.

In [8]:
reviews['condition'].value_counts()

Type 2 Diabetes Mellitus                     4955
Additional Medication for Diabetes Type 2     483
Diabetes                                       41
Type 1 Diabetes Mellitus                       16
Name: condition, dtype: int64

Ok, so 90% of our reviews are specifically for Type II Diabetes Mellitus, 9% on "Additional Medication for Type II"

In [9]:
reviews['reviewer'].nunique()

3916

Check for null values.

In [10]:
np.sum(reviews.isnull())

Rdate               0
comment          4520
condition           0
drug                0
easeofuse           0
effectiveness       0
helpful             0
reviewer          253
satisfaction        0
dtype: int64

Uh oh!  4,520 null comments out of 5,495 reviews...this can't be right. Let's grab examples and debug.

In [21]:
reviews.loc[reviews.isnull().any(axis=1), :]

Rdate                                  comment  \
1      9/18/2007 11:18:23 AM                                      NaN   
3        4/5/2011 1:31:52 PM                                      NaN   
4        3/7/2011 4:35:50 PM                                      NaN   
6      3/25/2015 12:19:46 PM                                      NaN   
7       3/24/2015 5:08:47 PM                                      NaN   
9     10/21/2007 10:23:33 AM                                      NaN   
10     10/14/2007 9:24:09 PM                                      NaN   
11     10/14/2007 8:59:09 PM                                      NaN   
13       5/7/2014 3:05:19 PM                                      NaN   
15     9/21/2007 10:48:32 AM                                  love it   
19     10/16/2007 4:33:20 PM  works fast and for long periods of time   
20      9/19/2007 5:10:22 PM                                      NaN   
24    12/20/2007 12:21:47 PM                                      NaN   
25     10/25/2007 7:33:33 PM                                      NaN   
26     10/21/2007 9:29:23 PM                                      NaN   
27      10/4/2007 9:19:04 PM                                      NaN   
29      3/26/2008 8:54:09 PM                                      NaN   
30      3/23/2008 1:32:09 AM                                      NaN   
31      3/16/2008 6:17:22 PM                                      NaN   
32     1/18/2008 12:45:07 PM                                      NaN   
34      6/18/2008 7:58:34 PM                                      NaN   
35       6/8/2008 7:38:08 AM                                      NaN   
36     5/31/2008 11:17:32 PM                                      NaN   
37      5/19/2008 7:47:38 PM                                      NaN   
39       3/1/2009 1:23:24 AM                                      NaN   
40     2/23/2009 10:52:17 PM                                      NaN   
41      1/28/2009 3:58:03 PM                                      NaN   
42      1/25/2009 7:26:50 PM                                      NaN   
43      6/1/2009 10:21:22 PM                                      NaN   
44      5/29/2009 6:51:14 AM                                      NaN   
...                      ...                                      ...   
5458    11/3/2016 1:01:03 PM                                      NaN   
5459   10/27/2016 9:04:40 AM                                      NaN   
5461   11/5/2017 12:19:10 AM                                      NaN   
5462  10/25/2017 11:31:09 PM                                      NaN   
5463    8/6/2017 12:06:43 PM                                      NaN   
5464    6/23/2017 1:58:23 PM                                      NaN   
5466   3/17/2019 12:37:36 PM                                      NaN   
5467     3/3/2019 7:28:22 PM                                      NaN   
5468    6/24/2018 4:03:32 PM                                      NaN   
5469   3/17/2018 10:00:48 AM                                      NaN   
5471   12/12/2013 5:07:20 PM                                      NaN   
5472    9/23/2013 5:43:21 PM                                      NaN   
5473    9/8/2013 10:11:24 AM                                      NaN   
5474   8/17/2013 11:14:47 AM                                      NaN   
5476   1/14/2014 10:30:21 PM                                      NaN   
5477     1/2/2014 8:10:15 AM                                      NaN   
5478   12/31/2013 1:24:53 AM                                      NaN   
5479  12/29/2013 11:14:50 AM                                      NaN   
5481    7/10/2014 5:06:02 PM                                      NaN   
5482    6/27/2014 8:22:44 AM                                      NaN   
5483    5/10/2014 2:20:01 PM                                      NaN   
5484     5/6/2014 5:31:23 PM                                      NaN   
5486     4/7/2015 1:11:37 PM                                      NaN   
5487    

Note: Identified bugs in review parser that caused nulls in 'comment' field. There ARE some blank comments on WebMD, but not on 80% of reviews! The 253 null 'reviewer's are genuinely blank fields on the website.

Debugged spider to fix review parsing.  Before re-crawling, let's take a look at ratings fields (below).

Now let's see some summary stats for our ratings!

In [11]:
reviews.describe()

easeofuse  effectiveness      helpful  satisfaction
count  5495.000000    5495.000000  5495.000000   5495.000000
mean      4.055505       3.442038     3.079709      3.052957
std       1.246186       1.425960     2.591082      1.596296
min       1.000000       1.000000     0.000000      1.000000
25%       3.000000       2.000000     1.000000      1.000000
50%       5.000000       4.000000     2.000000      3.000000
75%       5.000000       5.000000     5.000000      5.000000
max       5.000000       5.000000     9.000000      5.000000

Note: 'easeofuse', 'effectiveness', and 'satisfaction' fields are all 5-max ratings; however, 'helpful' is a counter where users upvote. '9' as a max isn't correct.  After reviewing site and finding examples from .loc table above, identified bug in code limiting to one integer. Fixed in code to allow more than 1 integer and rerun spider.

After debugging review parser for 'helpful' and 'comment' fields, re-crawling WebMD.

##### QA round 2:
...2 hours later...

Load 2nd-round file and QA again!

In [26]:
reviews = pd.read_csv('webmd_reviews2.csv')

In [27]:
reviews.sample(10)

Rdate  \
4320   2/5/2009 10:15:18 PM   
3687   6/4/2013 11:08:03 AM   
2332   9/21/2007 1:06:43 AM   
1210    9/2/2009 6:13:28 PM   
4218   4/14/2008 7:11:22 PM   
2120  5/17/2013 11:17:37 AM   
2373   1/8/2016 10:18:32 AM   
1408   7/12/2010 8:33:18 AM   
1732    4/7/2013 3:24:43 PM   
4984   4/16/2009 9:28:14 AM   

                                                comment  \
4320  I have been on Byetta for one week now, and si...   
3687  my type 2 diabetes has remained relatively unc...   
2332  does it cause gas..im having a gas problem..i ...   
1210  Good reactions -generally--\r\nvery responsive...   
4218  DX with Type 2 diabetes.  On Byetta for 16 mon...   
2120  I used this medicine for  three months.  My su...   
2373  Started Trulicity October 2015. A1C was 12 and...   
1408  It does help lower blood sugar, however, I get...   
1732  Started taking this in July of 12. 1000 mg twi...   
4984       no fuss-no muss-easey to use and no problems   

                                      condition                    drug  \
4320  Additional Medication for Diabetes Type 2     Byetta subcutaneous   
3687                   Type 2 Diabetes Mellitus              Actos oral   
2332                   Type 2 Diabetes Mellitus          Avandamet oral   
1210                   Type 2 Diabetes Mellitus          metformin oral   
4218  Additional Medication for Diabetes Type 2     Byetta subcutaneous   
2120                   Type 2 Diabetes Mellitus         Jentadueto oral   
2373                   Type 2 Diabetes Mellitus  Trulicity subcutaneous   
1408                   Type 2 Diabetes Mellitus          metformin oral   
1732                   Type 2 Diabetes Mellitus          metformin oral   
4984                   Type 2 Diabetes Mellitus          glipizide oral   

      easeofuse  effectiveness  helpful  \
4320          3              4        1   
3687          3              2       16   
2332          5              5       10   
1210          5              4        2   
4218          5              5       18   
2120          4              4        6   
2373          1              1       30   
1408          4              3        5   
1732          5              3       46   
4984          5              5        4   

                                               reviewer  satisfaction  
4320  masuccij, 45-54 Male  on Treatment for 5 to le...             3  
3687  james, 75 or over Male  on Treatment for 2 to ...             2  
2332  smichel6112, 55-64 Female  on Treatment for 2 ...             5  
1210  55-64 Female  on Treatment for 2 to less than ...             4  
4218  75 or over Female  on Treatment for 1 to 6 mon...             5  
2120  MommE2, 45-54 Female  on Treatment for less th...             3  
2373  Jltc, 55-64 Female  on Treatment for less than...             1  
1408  gego, 65-74 Female  on Treatment for 2 to less...             2  
1732                                                NaN             3  
4984  teachmjk, 45-54 Female  on Treatment for 1 to ...             5

In [28]:
np.sum(reviews.isnull())

Rdate              0
comment            9
condition          0
drug               0
easeofuse          0
effectiveness      0
helpful            0
reviewer         253
satisfaction       0
dtype: int64

In [29]:
reviews.describe()

easeofuse  effectiveness      helpful  satisfaction
count  5495.000000    5495.000000  5495.000000   5495.000000
mean      4.055505       3.442038     9.202002      3.052957
std       1.246186       1.425960     9.672916      1.596296
min       1.000000       1.000000     0.000000      1.000000
25%       3.000000       2.000000     3.000000      1.000000
50%       5.000000       4.000000     7.000000      3.000000
75%       5.000000       5.000000    13.000000      5.000000
max       5.000000       5.000000    79.000000      5.000000

This looks more like it!

#### Exploratory Data Analysis